Baikal/PostEffects/ML/sisr_preprocess.cpp

```diff
@@ -0,0 +1,149 @@
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+********************************************************************/
+
+#include "sisr_preprocess.h"
+#include "Output/clwoutput.h"
+#include "CLWBuffer.h"
+
+namespace Baikal
+{
+    namespace PostEffects
+    {
+        using uint32_t = std::uint32_t;
+        using float3 =  RadeonRays::float3;
+
+        SisrPreprocess::SisrPreprocess(CLWContext context,
+                                       Baikal::CLProgramManager const *program_manager,
+                                       std::uint32_t start_spp)
+        : DataPreprocess(context, program_manager, start_spp)
+        , m_context(mlCreateContext())
+        {}
+
+        void SisrPreprocess::Init(std::uint32_t width, std::uint32_t height)
+        {
+            m_cache = CLWBuffer<float3>::Create(GetContext(),
+                                                CL_MEM_READ_WRITE,
+                                                width * height);
+
+            m_input = CLWBuffer<float>::Create(GetContext(),
+                                               CL_MEM_READ_WRITE,
+                                               3 * width * height);
+
+            ml_image_info image_info = {ML_FLOAT32, width, height, 3};
+            m_image = mlCreateImage(m_context, &image_info);
+
+            if (!m_image)
+            {
+                throw std::runtime_error("can not create ml_image");
+            }
+        }
+
+        Image SisrPreprocess::MakeInput(PostEffect::InputSet const& inputs)
+        {
+            auto color_aov = inputs.begin()->second;
+
+            if (!m_is_init)
+            {
+                m_width = color_aov->width();
+                m_height = color_aov->height();
+                Init(m_width, m_height);
+                m_is_init = true;
+            }
+
+            auto clw_input = dynamic_cast<ClwOutput *>(color_aov);
+
+            if (clw_input == nullptr) {
+                throw std::runtime_error("SisrPreprocess::MakeInput(..): incorrect input");
+            }
+
+            auto context = GetContext();
+
+            // read spp from first pixel as 4th channel
+            RadeonRays::float3 pixel = .0f;
+            context.ReadBuffer<float3>(0, clw_input->data(), &pixel, 1).Wait();
+            auto sample_count = static_cast<unsigned>(pixel.w);
+
+            if (m_start_spp > sample_count)
+            {
+                return Image(sample_count, nullptr);
+            }
+
+            Tonemap(m_cache, clw_input->data());
+
+            // delete 4th channel
+            WriteToInputs(m_input,
+                          CLWBuffer<float>::CreateFromClBuffer(m_cache),
+                          m_width,
+                          m_height,
+                          0,  // dst channels offset
```

Comment:
I recommend `0 /*exact_parameter_name*/,`
- Exact name to make it easier to search.
- /* */ comment before the comma to avoid bad formatting by some IDEs.